flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/ExpressionConverter.java

```diff
@@ -235,27 +255,96 @@ public RexNode visit(TypeLiteralExpression typeLiteral) {
     }
 
     @Override
-    public RexNode visit(Expression other) {
+    public RexNode visit(TableReferenceExpression tableRef) {
+        final LogicalType tableArgType = tableRef.getOutputDataType().getLogicalType();
+        final RelDataType rowType = typeFactory.buildRelNodeRowType((RowType) tableArgType);
+        final int[] partitionKeys;
+        if (tableRef.getQueryOperation() instanceof PartitionQueryOperation) {
+            final PartitionQueryOperation partitionOperation =
+                    (PartitionQueryOperation) tableRef.getQueryOperation();
+            partitionKeys = partitionOperation.getPartitionKeys();
+        } else {
+            partitionKeys = new int[0];
+        }
+        final RexTableArgCall tableArgCall =
+                new RexTableArgCall(rowType, inputStack.size(), partitionKeys, new int[0]);
+        inputStack.add(relBuilder.build());
```

Comment:
Can we maintain the stack and do the relBuilder.build()  in the QueryOperationConverter?

Reply:
Looks we still need to special handle `RexTableArgCall` here, how about we leave original logic for `TableReferenceExpression` and just move `ModelReferenceExpression` to `ExpressionConverter`?

Reply:
Created https://issues.apache.org/jira/browse/FLINK-38638